community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -19,112 +19,302 @@
+        unit.next = stack.getAndSet( unit ); // benign race, see reverse()
+        return unit;
+    }
+
+    private Throwable tryDoWork( WorkUnit<Material,W> unit, int tryCount, boolean otherwisePark )
+    {
+        if ( tryLock( tryCount, unit, otherwisePark ) )
+        {
+            WorkUnit<Material,W> batch = null;
+            try
+            {
+                batch = reverseAndFilterCancelled( stack.getAndSet( stackEnd ) );
+                return doSynchronizedWork( batch );
+            }
+            finally
+            {
+                unlock();
+                unparkAnyWaiters();
+                markAsDone( batch );
+            }
+        }
+        return null;
+    }
+
+    private void unparkAnyWaiters()
+    {
+        WorkUnit<Material,W> nextBatch = stack.get();
+        while ( nextBatch != stackEnd )
+        {
+            if ( !nextBatch.isCancelled() )
+            {
+                nextBatch.unpark();
+                break;
+            }
+            WorkUnit<Material,W> next;
+            do
+            {
+                next = nextBatch.next;
+            }
+            while ( next == null );
+            nextBatch = next;
+        }
+    }
+
+    private void checkFailure( Throwable failure ) throws ExecutionException
+    {
+        if ( failure != null )
+        {
+            throw new ExecutionException( failure );
+        }
+    }
+
+    private boolean tryLock( int tryCount, WorkUnit<Material,W> unit, boolean otherwisePark )
+    {
+        if ( lock.compareAndSet( null, Thread.currentThread() ) )
+        {
+            // Got the lock!
+            return true;
+        }
+
+        // Did not get the lock, spend some time until our work has aither been completed,
+        // or we get the lock.
+        if ( otherwisePark )
+        {
+            if ( tryCount < 1000 )
+            {
+                // todo Thread.onSpinWait() in Java9?
+                Thread.yield();
             }
-            catch ( InterruptedException e )
+            else
             {
-                // We can't stop now, because our work has already been
-                // scheduled. So instead we're just going to reset the
-                // interruption status when we're done.
-                wasInterrupted = true;
+                unit.park( 10, TimeUnit.MILLISECONDS );
             }
         }
-        while ( !unit.done );
+        return false;
+    }
 
-        if ( wasInterrupted )
+    private void unlock()
+    {
+        if ( lock.getAndSet( null ) != Thread.currentThread() )
         {
-            Thread.currentThread().interrupt();
+            throw new IllegalMonitorStateException(
+                    "WorkSync accidentally released a lock not owned by the current thread" );
         }
     }
 
-    private void doSynchronizedWork()
+    private Throwable doSynchronizedWork( WorkUnit<Material,W> batch )
```

Comment:
the name `doSynchronizedWork` is a bit confusing, since the work isn't _synchronized_ in any way. I understand that the name is supposed to signal that at this point there is only one thread doing work, but perhaps we can find a better way of signalling that? How about just simply calling the method `doWork`, and add a javadoc comment about the fact that the locking in `tryDoWork` ensures that there is only one thread performing work here at any given time?